### PR TITLE
[stdlib] Improve performance of heterogeneous binary integer `==` and `<`

### LINKED ISSUE #62260: Heterogeneous integer comparisons produce suboptimal code
**Description**

Consider two Swift functions, both comparing an `Int` with a `UInt32`:

```swift
func compare_direct(_ x: Int, _ y: UInt32) -> Bool {
    x < y
}

func compare_widened(_ x: Int, _ y: UInt32) -> Bool {
    x < Int(y)
}
``` 

These compile to vastly different code.

The versions which widens `y` compiles to a simple `cmp` instruction:

```asm
output.compare_widened(Swift.Int, Swift.UInt32) -> Swift.Bool:
        mov     eax, esi
        cmp     rax, rdi
        setg    al
        ret
```

But the version which performs the comparison directly compiles to a whole algorithm:

```asm
output.compare_direct(Swift.Int, Swift.UInt32) -> Swift.Bool:
        test    rdi, rdi
        js      .LBB1_1
        jne     .LBB1_6
        test    esi, esi
        jne     .LBB1_6
        xor     eax, eax
        ret
.LBB1_1:
        mov     al, 1
        ret
.LBB1_6:
        mov     eax, esi
        cmp     rax, rdi
        seta    al
        ret
```

**Steps to reproduce**

[Godbolt](https://swift.godbolt.org/#g:!((g:!((g:!((h:codeEditor,i:(filename:'1',fontScale:14,fontUsePx:'0',j:1,lang:swift,selection:(endColumn:2,endLineNumber:7,positionColumn:2,positionLineNumber:7,selectionStartColumn:2,selectionStartLineNumber:7,startColumn:2,startLineNumber:7),source:'func+compare_direct(_+x:+Int,+_+y:+UInt32)+-%3E+Bool+%7B%0A++++x+%3C+y%0A%7D%0A%0Afunc+compare_widened(_+x:+Int,+_+y:+UInt32)+-%3E+Bool+%7B%0A++++x+%3C+Int(y)%0A%7D'),l:'5',n:'0',o:'Swift+source+%231',t:'0')),k:54.577893820713676,l:'4',n:'0',o:'',s:0,t:'0'),(g:!((h:compiler,i:(compiler:swiftnightly,deviceViewOpen:'1',filters:(b:'0',binary:'1',commentOnly:'0',demangle:'0',directives:'0',execute:'1',intel:'0',libraryCode:'0',trim:'1'),flagsViewOpen:'1',fontScale:14,fontUsePx:'0',j:1,lang:swift,libs:!(),options:'-O',selection:(endColumn:12,endLineNumber:20,positionColumn:1,positionLineNumber:5,selectionStartColumn:12,selectionStartLineNumber:20,startColumn:1,startLineNumber:5),source:1),l:'5',n:'0',o:'+x86-64+swiftc+nightly+(Editor+%231)',t:'0')),k:45.42210617928634,l:'4',n:'0',o:'',s:0,t:'0')),l:'2',n:'0',o:'',t:'0')),version:4)

**Expected behavior**

These functions do the same job (I think -- always a dangerous assumption to make with numerics 😅); they should compile to the same code. The single-`cmp` variant would appear to be the optimal solution.

**Environment**

- Swift compiler version info

```
Swift version 5.7 (swift-5.7-RELEASE)
Target: x86_64-unknown-linux-gnu
```

### PR DESCRIPTION
**Pull request by xwu**
As it turns out, we can simplify this logic quite a bit and get more optimal generated code in the process:

In the first iteration of this PR, we unconditionally apply the "roundtripping" logic used in the existing code to widen a comparand without checking the instance property `bitWidth`:

- If `rhs` can be represented as a value of type `Self`, or if `lhs` can be represented as a value of type `Other`, then the comparison is straightforward.
- If neither is the case, then we can actually know which of `lhs` and `rhs` is smaller just by knowing which of `Self` or `Other` is a signed type (see comments in the code).

This change allows Swift to generate the same, optimal code (at `-O`) for the following input as reported in #62260, which isn't the case today:

```swift
func compare_direct(_ x: Int, _ y: UInt32) -> Bool {
    x < y
}

func compare_widened(_ x: Int, _ y: UInt32) -> Bool {
    x < Int(y)
}
```

However, because the roundtripping logic has to start with either the left-hand side or the right-hand side, the same optimal output can't be obtained when we swap the parameters above—i.e., when `x` is of type `UInt32` and `y` is of type `Int`. The generated code is still _better_ than the status quo, but it's not the _best possible_.

---

To address this issue, in the second iteration of this PR, we compare comparand bit widths explicitly, but only once along each code path in case it's expensive to compute for values of a hypothetical arbitrary-width type. (I'd expect this should be negligible as compared to the widening operation that follows anyway.)

For fixed-width integer types, this approach produces _optimal_ code generation for heterogeneous comparison of any two types. This is because we defer comparison to zero until we reach the only scenario where it's required (when a signed and therefore possibly negative value is being compared to an unsigned value of greater bit width) and otherwise stick to operations that can be optimized away in the concrete case.

This change apparently speeds up generic floating-point conversion severalfold, improves code size for a number of integer and floating-point algorithms, and even shrinks `libswiftCore.dylib` code size by 2%. It also revealed brittleness in an IRGen test which has been modified as a result to improve robustness.

Resolves #62260.

### CODE REVIEW

**Comment by xwu** [comment]
Would appreciate guidance on where best to put a test for this code generation.

**Comment by xwu** [comment]
@swift-ci test

**Comment by xwu** [comment]
@swift-ci benchmark

**Comment by xwu** [comment]
> In the implementation comments, I'm not sure that "bitcasting" or "bitcast operation" are the correct terms.

What would you suggest? I’d like something more succinct than “constructing a value of `Foo` type from
the bit pattern obtained by truncating or sign-extending as necessary the bit pattern of `bar`.” To do so, I’m adapting the terminology used in the existing implementation comment but I’m certainly open to other suggestions.

> Could the `(0 as Self)` and `(0 as Other)` expressions be replaced by `Self.zero` and `Other.zero`?

It could, just as most if not all uses of literal `0` could be replaced by `.zero`, but we’ve always said that `0` is the preferred canonical spelling for integers. That there is an explicit type here is a reflection of an extant bug that causes `<` in generic contexts (but not concrete contexts) not to favor homogeneous comparison; this ought to be fixed because it is a footgun, and at such point when it is, this code can be rewritten `lhs < 0`, dropping the explicit type coercion.

I’d rather the workaround until then be (a) visually thorny rather than potentially appear like a stylistic preference; (b) generally applicable for all values—we can’t very well write `lhs == Self.fortyTwo`—so that authors who need to work on this code can see both the need for this workaround throughout the implementation and how to apply it no matter the comparand.

**Comment by xwu** [comment]
@swift-ci test

**Comment by xwu** [comment]
@swift-ci benchmark

**Comment by xwu** [comment]
Irksome is the apparent benchmark regression in string–substring comparison in all optimization modes; not sure how to explain that one.

This PR also apparently affects the temporary allocation codegen test, which I'll need to update I guess.

**Comment by xwu** [comment]
@swift-ci test Linux platform

**Comment by xwu** [comment]
@swift-ci test macOS platform

**Comment by xwu** [comment]
> Most of the benchmark regressions are marked with (?) — "false alarms" or "noise". Can any of the results be trusted?

The string comparison benchmarks are consistently showing up, so it'll take some manual inspection of the generated code to know if it's noise or real.

**Comment by xwu** [comment]
@swift-ci please build toolchain

**Comment by xwu** [comment]
Can confirm identical SIL is generated for all benchmarks in `SubstringTest` using the toolchain versus the nightly, which answers that concern.

**Comment by xwu** [comment]
@swift-ci test Linux platform

**Comment by xwu** [comment]
@swift-ci test Linux platform

**Comment by xwu** [comment]
@swift-ci test

**Comment by xwu** [comment]
@swift-ci test

**Comment by xwu** [comment]
@swift-ci test

**Comment by xwu** [comment]
@swift-ci test

**Comment by xwu** [comment]
@swift-ci test macOS platform

**Comment by xwu** [comment]
@grynspan I've had to make edits to the temporary allocation codegen test to keep CI happy.

In the IR, the relative order in which basic blocks for stack versus heap allocation are emitted for the large allocation test is arbitrary, but the existing test only accepts one specific order. With this PR, the basic blocks are now omitted in a different order on macOS because the generated code (rightly) has some different inlining choices earlier in the file with respect to a comparison operation with a randomly generated fixed-width integer.

To make the test robust to arbitrary changes in the order of basic blocks, I've changed certain "CHECK"s to "CHECK-DAG"s. At some point in the past decade, folks proposed a "CHECK-LABEL-DAG" feature to be added to FileCheck, but unfortunately it didn't gain traction and it doesn't appear possible to checkentire basic blocks being reordered relative to one another in one invocation of FileCheck, just single lines. Therefore, I've also changed the test to invoke FileCheck three times, with the first run testing for presence of the expected branch instructions, and the latter two specifically testing either the heap allocation IR or the stack allocation IR.

Note that the test as it currently exists on `main` has a RUN line that actually only tells FileCheck to check the platform-specific lines, as far as I can tell, so the vast majority of the "CHECK"s aren't being tested; I've changed the flags on the first invocation so that all "CHECK"s are actually tested--and fortunately it all seems to work fine.

Let me know if this change looks at least reasonable to you. I'd like to defer any detailed tinkering to you or others as it's not salient to this PR, but some degree of change was required to unblock this work and I figured I'd at least make a real effort.

**Review by grynspan — APPROVED** [review]
I don't object to the changes to the temp-alloc unit tests, so long as they continue to pass. ;)

**Comment by stephentyrone** [comment]
I'm planning to rework these further, but this seems like a reasonable start...

File: stdlib/public/core/Integers.swift [hunk 1]
```diff
@@ -1730,34 +1717,26 @@ extension BinaryInteger {
...
-    return Other(truncatingIfNeeded: lhs) < rhs
+    // If `Self` is signed but `Other` is unsigned, then we have to
+    // be a little more careful about widening, since:
+    // (1) a fixed-width signed type can't represent the largest values of
+    //     a fixed-width unsigned type of equal bit width; and
+    // (2) an unsigned type (obviously) can't represent a negative value.
+    if Self.isSigned {
+      return lhs.bitWidth > rhs.bitWidth ? // (1)
+        lhs < Self(truncatingIfNeeded: rhs) :
+        lhs < (0 as Self) ? true : Other(truncatingIfNeeded: lhs) < rhs // (2)
```
Comment (stephentyrone) [inline stdlib/public/core/Integers.swift:1734]: I would prefer to write this as `lhs < (0 as Self) || Other(...)` unless that introduces some problem.

**Comment by stephentyrone** [comment]
One style nit, otherwise LGTM.

**Comment by xwu** [comment]
@swift-ci test and merge

**Comment by shahmishal** [comment]
We are seeing multiple aarch64 bots failing: 

```
Failed Tests (1):
  Swift(linux-aarch64) :: IRGen/temporary_allocation/codegen.swift
```

https://ci.swift.org/job/oss-swift-package-ubuntu-20_04-aarch64//1442/console

cc: @xwu @stephentyrone

**Comment by xwu** [comment]
🫤 this is what I get for trying to unblock things by fixing rather than just disabling the test.

Most of the tests weren't actually being checked and my good deed was properly enabling it, but looks like `zeroext` on that call is applicable only for x86_64; I think the best way to unblock is to disable this test again for arm64.

cc @grynspan

**Comment by xwu** [comment]
#63093

**Comment by xwu** [comment]
Incidentally, @shahmishal, given CI passed multiple times, was there any way to spot this issue pre-emptively?

**Pull request merged**